[Visual Bidi Selection] Set selection to visually- and logically-contiguous boundaries when ending range adjustment

whsieh opened the pull request:
#### e59afa8f7cc66431f54539e0607a1f2948c8fb89
<pre>
[Visual Bidi Selection] Set selection to visually- and logically-contiguous boundaries when ending range adjustment
<a href="https://bugs.webkit.org/show_bug.cgi?id=283515">https://bugs.webkit.org/show_bug.cgi?id=283515</a>
<a href="https://rdar.apple.com/140378287">rdar://140378287</a>

Reviewed by Ryosuke Niwa and Abrar Rahman Protyasha.

Work towards visual bidi selection support, by limiting visually contiguous bidi selection behavior
to only when the user is selecting text with selection handles on iOS (i.e. range adjustment). When
range adjustment ends, we then snap the selection endpoints to the nearest bidi text run boundaries
that ensure both logical and visual contiguity. See below for more details.

* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-expected.txt:
* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection.html:

Augment an existing layout test to exercise the selection snapping behavior, by verifying that the
final selection is snapped to the start of the Arabic text within the (otherwise English) paragraph.

* Source/WebCore/editing/Editing.cpp:
(WebCore::visualDistanceOnSameLine):

Add a helper function to compute the number of characters (visually) between two `RenderedPosition`s
on the same line box.

(WebCore::visuallyClosestBidiBoundary):

Add a helper function to return the (visually) closest bidi boundary point around a given
`RenderedPosition`, at the given bidi level.

(WebCore::adjustToVisuallyContiguousRange):

Implement the core of the bidi selection snapping heuristic. This new editing helper function,
`adjustToVisuallyContiguousRange`, takes a `SimpleRange` and returns another `SimpleRange` that is
either expanded or contracted to maintain both visual and logical contiguity for bidi text.

This algorithm generally works by determining whether or not the start or end positions (separately)
need to be adjusted to the nearest bidi boundary in order to maintain visual and logical contiguity.
To achieve this, we first examine the line boxes in between the start and end positions for a
single-line selection, or alternately, between the &apos;selection start and logical end on the first
line&apos; and &apos;selection end and logical start on the last line&apos; for a multi-line selection. During this
traversal, we keep track of the minimum observed bidi level for any line box we traverse over — if
a selection endpoint has a bidi level exceeding this minimum bidi level, then it&apos;ll result in a
visual discontiguity.

Consider, for example, the case of selection endpoints within two nested bidi text runs with the
same bidi level (the selection starts in `bbbbb` and ends in `ddddd`). The entire text in logical
order is `aaaaa bbbbb ccccc ddddd`:

DIRECTION    ltr   rtl   ltr   rtl
            ----&gt; &lt;---- ----&gt; &lt;----
LEVEL         0     1     2     1
TEXT        aaaaa ddddd ccccc bbbbb
                    ^           ^
SELECTION          End        Start

Because the bidi level in between the start and end only *increases*, the selection is already
visually contiguous, so no endpoint adjustment is needed despite the fact that the selection spans
a combination of RTL and LTR text. This is because any bidi text that starts within the selected
range also ends in the same selected range. However, consider this other, vaguely similar-looking
scenario:

DIRECTION    ltr   rtl   ltr   rtl
            ----&gt; &lt;---- ----&gt; &lt;----
LEVEL         0     1     0     1
TEXT        aaaaa bbbbb ccccc ddddd
                    ^           ^
SELECTION         Start        End

Because the selection now encompasses a piece of text with a *lower* bidi level, both endpoints must
be adjusted to their respective boundaries in order to maintain both visual and logical contiguity;
the start must either move to the end of `aaaaa` or the beginning of `ccccc`, and the end must move
to the end of `ccccc` or the end of `ddddd`. This is because this second selected range contains a
run of text that logically continues before the start or after the end (in this example, before the
start).

For each selection endpoint that needs to be adjusted in order to land on a visually contiguous bidi
boundary, we then use the helper methods in `RenderedPosition` to identify the nearest 2 bidi
boundary points on the same line around the endpoint, and select the one that results in the least
amount of (visual) adjustment.

* Source/WebCore/editing/Editing.h:
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/FrameSelection.h:

Drive-by fix: delete this unused method declaration.

* Source/WebCore/editing/RenderedPosition.cpp:
(WebCore::RenderedPosition::leftBoundaryOfBidiRun const):
(WebCore::RenderedPosition::rightBoundaryOfBidiRun const):
(WebCore::RenderedPosition::leftBoundaryOfBidiRun): Deleted.
(WebCore::RenderedPosition::rightBoundaryOfBidiRun): Deleted.
* Source/WebCore/editing/RenderedPosition.h:
(WebCore::RenderedPosition::box const):
(WebCore::RenderedPosition::offset const):

Expose a couple of new getters on `RenderedPosition`, for use in the new helper function in
`Editing.cpp`.

* Source/WebCore/editing/ios/EditorIOS.mm:
(WebCore::Editor::shouldDrawVisuallyContiguousBidiSelection const):
* Source/WebCore/page/EditorClient.h:
(WebCore::EditorClient::shouldDrawVisuallyContiguousBidiSelection const):
* Source/WebCore/platform/ios/SelectionGeometry.cpp:
(WebCore::SelectionGeometry::SelectionGeometry):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/ios/SelectionGeometry.h:
(WebCore::SelectionGeometry::behavior const):
(WebCore::SelectionGeometry::setBehavior):
(WebCore::SelectionGeometry::mayAppearLogicallyDiscontiguous const): Deleted.
(WebCore::SelectionGeometry::setMayAppearLogicallyDiscontiguous): Deleted.

Remove this incorrect logic; see below for more details.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::makeBidiSelectionVisuallyContiguousIfNeeded):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.mm:
(-[WKTextSelectionRect writingDirection]):

Revert this hack to force LTR selection behavior when visually contiguous bidi selection is enabled.
This is incorrect because it doesn&apos;t take line direction into account in RTL paragraphs — in a
future change, I&apos;ll fix this properly by correctly setting the text direction of coalesced selection
geometries, based on the primary direction on each line.

* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/ios/WebEditorClientIOS.mm:
(WebKit::WebEditorClient::shouldDrawVisuallyContiguousBidiSelection const):

Add an editor client hook to determine whether or not we should draw visually contiguous bidi
selections; we currently only do this when the user is adjusting selection handles on iOS, but in a
future patch I&apos;ll update this to include more text interactions, such as tap-and-half, loupe, and
selecting with a mouse cursor using trackpad.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::setShouldDrawVisuallyContiguousBidiSelection):
(WebKit::WebPage::updateSelectionWithTouches):

Canonical link: <a href="https://commits.webkit.org/287021@main">https://commits.webkit.org/287021@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d63eb2029fc5a57e6a7c8780755b5105e126f3d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77972 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57007 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31347 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82622 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29230 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80095 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66169 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5300 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61061 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18989 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81040 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51053 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/67319 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41363 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48416 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/24508 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27573 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69509 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24855 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83984 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5339 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3613 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69283 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5495 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66905 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68538 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17113 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12536 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10714 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5287 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/8040 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5279 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8711 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7064 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/editing/Editing.cpp:
```diff
@@ -1222,4 +1223,117 @@ HashSet<RefPtr<HTMLImageElement>> visibleImageElementsInRangeWithNonLoadedImages
+        unsigned maxOffset { 0 };
+        TextDirection direction { TextDirection::LTR };
+
+        unsigned totalLength() const { return maxOffset > minOffset ? maxOffset - minOffset : 0; }
```
Maybe `std::max(maxOffset - minOffset, 0)`?

aprotyas replied:
oh no, unsigned arithmetic 😞

aprotyas replied:
Wired: `unsigned totalLength() const { return std::max(maxOffset, minOffset) - minOffset; }`

whsieh replied:
SGTM, I'll go with the last approach (for `totalLength` and `visuallySelectedLength`)

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -1936,12 +1936,41 @@ static bool insideImageOverlay(const VisiblePosition& position)
+    m_shouldDrawVisuallyContiguousBidiSelection = value;
+
+    if (!value)
+        scheduleFullEditorStateUpdate();
```
Why is an editor state update not needed when going false->true?

whsieh replied:
Good catch — it's (technically) not needed at the moment because whenever we enter this mode, we're about to start changing the selection anyways which will schedule editor state updates.

However, I agree it makes sense to `scheduleFullEditorStateUpdate();` here, since we're not *always* guaranteed to trigger subsequent ES updates in the future. I'll remove the `!value` check here.

rniwa commented on Source/WebCore/editing/Editing.cpp:
```diff
@@ -1222,4 +1223,117 @@ HashSet<RefPtr<HTMLImageElement>> visibleImageElementsInRangeWithNonLoadedImages
+    };
+
+    Vector<VisuallyAdjacentTextRange> ranges;
+    for (auto box = fromBox; box; box = advance(box)) {
```
So a while ago, I added RenderedPosition to deal with walk inline boxes accounting for bidi-levels.
Maybe we can use it here?

whsieh replied:
Good call — yes, I think I'll need to adopt `RenderedPosition` here, especially so that I can achieve some sensible behavior with respect to nested bidi. Will try out a few things and update the PR — thanks!

rniwa commented on Source/WebCore/editing/Editing.cpp:
```diff
@@ -1222,4 +1224,130 @@ HashSet<RefPtr<HTMLImageElement>> visibleImageElementsInRangeWithNonLoadedImages
+    if (first.box() == second.box())
+        return std::max(first.offset(), second.offset()) - std::min(first.offset(), second.offset());
+
+    auto distanceToBoundary = [](const RenderedPosition& position, bool leftSide) {
```
"leftSide" isn't always left side, is it?

whsieh replied:
Ah, so for this function, the `leftSide` parameter represents whether to count from the visual left boundary of the text run, or the visual right boundary. The `leftSide == (box->direction() == TextDirection::LTR)` check below then accounts for the difference in text direction (e.g., "the distance from the left boundary to the position offset in RTL text is computed by subtracting the (logical) max offset from the position offset").

To help clarify this, I'll refactor the code:
- Rename `distanceToBoundary` → `distanceFromOffsetToVisualBoundary`
- Use `enum class VisualBoundary` with values `Left/Right` instead of a `bool`.

Thanks for the review!

whsieh committed:
[Visual Bidi Selection] Set selection to visually- and logically-contiguous boundaries when ending range adjustment
https://bugs.webkit.org/show_bug.cgi?id=283515
rdar://140378287

Reviewed by Ryosuke Niwa and Abrar Rahman Protyasha.

Work towards visual bidi selection support, by limiting visually contiguous bidi selection behavior
to only when the user is selecting text with selection handles on iOS (i.e. range adjustment). When
range adjustment ends, we then snap the selection endpoints to the nearest bidi text run boundaries
that ensure both logical and visual contiguity. See below for more details.

* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-expected.txt:
* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection.html:

Augment an existing layout test to exercise the selection snapping behavior, by verifying that the
final selection is snapped to the start of the Arabic text within the (otherwise English) paragraph.

* Source/WebCore/editing/Editing.cpp:
(WebCore::visualDistanceOnSameLine):

Add a helper function to compute the number of characters (visually) between two `RenderedPosition`s
on the same line box.

(WebCore::visuallyClosestBidiBoundary):

Add a helper function to return the (visually) closest bidi boundary point around a given
`RenderedPosition`, at the given bidi level.

(WebCore::adjustToVisuallyContiguousRange):

Implement the core of the bidi selection snapping heuristic. This new editing helper function,
`adjustToVisuallyContiguousRange`, takes a `SimpleRange` and returns another `SimpleRange` that is
either expanded or contracted to maintain both visual and logical contiguity for bidi text.

This algorithm generally works by determining whether or not the start or end positions (separately)
need to be adjusted to the nearest bidi boundary in order to maintain visual and logical contiguity.
To achieve this, we first examine the line boxes in between the start and end positions for a
single-line selection, or alternately, between the 'selection start and logical end on the first
line' and 'selection end and logical start on the last line' for a multi-line selection. During this
traversal, we keep track of the minimum observed bidi level for any line box we traverse over — if
a selection endpoint has a bidi level exceeding this minimum bidi level, then it'll result in a
visual discontiguity.

Consider, for example, the case of selection endpoints within two nested bidi text runs with the
same bidi level (the selection starts in `bbbbb` and ends in `ddddd`). The entire text in logical
order is `aaaaa bbbbb ccccc ddddd`:

DIRECTION    ltr   rtl   ltr   rtl
            ----> <---- ----> <----
LEVEL         0     1     2     1
TEXT        aaaaa ddddd ccccc bbbbb
                    ^           ^
SELECTION          End        Start

Because the bidi level in between the start and end only *increases*, the selection is already
visually contiguous, so no endpoint adjustment is needed despite the fact that the selection spans
a combination of RTL and LTR text. This is because any bidi text that starts within the selected
range also ends in the same selected range. However, consider this other, vaguely similar-looking
scenario:

DIRECTION    ltr   rtl   ltr   rtl
            ----> <---- ----> <----
LEVEL         0     1     0     1
TEXT        aaaaa bbbbb ccccc ddddd
                    ^           ^
SELECTION         Start        End

Because the selection now encompasses a piece of text with a *lower* bidi level, both endpoints must
be adjusted to their respective boundaries in order to maintain both visual and logical contiguity;
the start must either move to the end of `aaaaa` or the beginning of `ccccc`, and the end must move
to the end of `ccccc` or the end of `ddddd`. This is because this second selected range contains a
run of text that logically continues before the start or after the end (in this example, before the
start).

For each selection endpoint that needs to be adjusted in order to land on a visually contiguous bidi
boundary, we then use the helper methods in `RenderedPosition` to identify the nearest 2 bidi
boundary points on the same line around the endpoint, and select the one that results in the least
amount of (visual) adjustment.

* Source/WebCore/editing/Editing.h:
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/FrameSelection.h:

Drive-by fix: delete this unused method declaration.

* Source/WebCore/editing/RenderedPosition.cpp:
(WebCore::RenderedPosition::leftBoundaryOfBidiRun const):
(WebCore::RenderedPosition::rightBoundaryOfBidiRun const):
(WebCore::RenderedPosition::leftBoundaryOfBidiRun): Deleted.
(WebCore::RenderedPosition::rightBoundaryOfBidiRun): Deleted.
* Source/WebCore/editing/RenderedPosition.h:
(WebCore::RenderedPosition::box const):
(WebCore::RenderedPosition::offset const):

Expose a couple of new getters on `RenderedPosition`, for use in the new helper function in
`Editing.cpp`.

* Source/WebCore/editing/ios/EditorIOS.mm:
(WebCore::Editor::shouldDrawVisuallyContiguousBidiSelection const):
* Source/WebCore/page/EditorClient.h:
(WebCore::EditorClient::shouldDrawVisuallyContiguousBidiSelection const):
* Source/WebCore/platform/ios/SelectionGeometry.cpp:
(WebCore::SelectionGeometry::SelectionGeometry):
(WebCore::operator<<):
* Source/WebCore/platform/ios/SelectionGeometry.h:
(WebCore::SelectionGeometry::behavior const):
(WebCore::SelectionGeometry::setBehavior):
(WebCore::SelectionGeometry::mayAppearLogicallyDiscontiguous const): Deleted.
(WebCore::SelectionGeometry::setMayAppearLogicallyDiscontiguous): Deleted.

Remove this incorrect logic; see below for more details.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::makeBidiSelectionVisuallyContiguousIfNeeded):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.mm:
(-[WKTextSelectionRect writingDirection]):

Revert this hack to force LTR selection behavior when visually contiguous bidi selection is enabled.
This is incorrect because it doesn't take line direction into account in RTL paragraphs — in a
future change, I'll fix this properly by correctly setting the text direction of coalesced selection
geometries, based on the primary direction on each line.

* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/ios/WebEditorClientIOS.mm:
(WebKit::WebEditorClient::shouldDrawVisuallyContiguousBidiSelection const):

Add an editor client hook to determine whether or not we should draw visually contiguous bidi
selections; we currently only do this when the user is adjusting selection handles on iOS, but in a
future patch I'll update this to include more text interactions, such as tap-and-half, loupe, and
selecting with a mouse cursor using trackpad.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::setShouldDrawVisuallyContiguousBidiSelection):
(WebKit::WebPage::updateSelectionWithTouches):

Canonical link: https://commits.webkit.org/287021@main
